### PR TITLE
Jonatan/mon 2301 improved model data alerts

### PR DESCRIPTION
**Pull request by JKL98ISR**

### CODE REVIEW

File: backend/deepchecks_monitoring/api/v1/configuration.py [hunk 1]
```diff
@@ -151,7 +171,81 @@ async def get_all_alert_rules(
...
+        alert_rule_schema = DataAlertRuleConfigSchema.from_orm(row)
+        if row.created_by != 0:
+            q = select(User).where(User.id == row.created_by)
+            user = (await session.execute(q)).scalars().first()
+            if user is not None:
+                user_schema = BasicUserSchema.from_orm(user)
+                alert_rule_schema.user = user_schema
+        data_alert_rule_schemas.append(alert_rule_schema)
+
+    return AlertRulesConfigSchema(alert_rules=alert_rule_schemas, data_alert_rules=data_alert_rule_schemas)
```
Comment (yromanyshyn) [inline backend/deepchecks_monitoring/api/v1/configuration.py:248]: IMO it is better to have two separate endpoints instead of one/
execution time of this one I think will be very big

File: backend/deepchecks_monitoring/api/v1/configuration.py [hunk 2]
```diff
@@ -151,7 +171,81 @@ async def get_all_alert_rules(
...
+        if "alert-window:desc" in sortby:
+            q = q.order_by(total_count.c.recent_alert.desc())
+
+    alert_rules_rows = (await session.execute(q)).all()
+    data_alert_rule_schemas = []
+    for row in alert_rules_rows:
+        alert_rule_schema = DataAlertRuleConfigSchema.from_orm(row)
+        if row.created_by != 0:
+            q = select(User).where(User.id == row.created_by)
+            user = (await session.execute(q)).scalars().first()
```
Comment (yromanyshyn) [inline backend/deepchecks_monitoring/api/v1/configuration.py:242]: issuing SQL queries within a loop in my opinion is not a very good idea, 
Did you try to `join` rules creators instead of selecting them sequentially?
which solution will have a better performance? (maybe create a separate issue for this)

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+from deepchecks_monitoring.schema_models.data_ingestion_alert_rule import DataIngestionAlertRule
+from deepchecks_monitoring.schema_models.model import Model
+from deepchecks_monitoring.schema_models.model_memeber import ModelMember
+from deepchecks_monitoring.utils import auth
+from deepchecks_monitoring.utils.alerts import AlertSeverity
+
+from .router import router
+
+
+class DataIngestionAlertCreationSchema(BaseModel):
```
Comment (yromanyshyn) [inline backend/deepchecks_monitoring/api/v1/data_ingestion_alert.py:32]: Why do you need a creation schema? Are clients allowed to create alerts manually?

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                    .where(ModelMember.user_id == user.id)
+                    .where(DataIngestionAlert.resolved == false()))
+    q = select_alert.group_by(DataIngestionAlertRule.alert_severity)
+    results = await session.execute(q)
+    total = results.all()
+    return dict(total)
+
+
+@router.post("/data-ingestion-alerts/{data_ingestion_alert_id}/resolve",
+             dependencies=[Depends(DataIngestionAlert.get_object_from_http_request)],
```
Comment (yromanyshyn) [inline backend/deepchecks_monitoring/api/v1/data_ingestion_alert.py:74]: why here, why not as a parameter?

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    name: t.Optional[str]
+    alert_severity: t.Optional[AlertSeverity]
+    condition: t.Optional[Condition]
+    is_active: t.Optional[bool]
+
+
+@router.post(
+    "/models/{model_id}/data-ingestion-alert-rules",
+    response_model=IdResponse,
+    tags=[Tags.ALERTS],
```
Comment (yromanyshyn) [inline backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py:71]: maybe use another tag for data ingestion alerts endpoints?
then it will be easier for the frontend team to find a required endpoint spec

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    dependencies=[Depends(Model.get_object_from_http_request)],
```
Comment (yromanyshyn) [inline backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py:72]: `Model.get_object_from_http_request`

I think we should decouple this into `Model.exists_or_404` and `Model.has_permission_or_403`/
I know that this is not related to this issue, so, maybe create an issue for this? 
of course, if you think that it has a sense

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            dependencies=[Depends(DataIngestionAlertRule.get_object_from_http_request)],
+            summary="Update data-ingestion alert rule by id.")
+async def update_alert(
+        data_ingestion_alert_rule_id: int,
+        body: DataIngestionAlertRuleUpdateSchema,
+        session: AsyncSession = AsyncSessionDep,
+        user: User = Depends(auth.CurrentUser()),
+):
+    """Update data-ingestion alert rule by id."""
+    updated_body = body.dict(exclude_unset=True).copy()
```
Comment (yromanyshyn) [inline backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py:115]: I think you also need to add `exclude_defaults`/

example

```python
# %%
import pydantic

class Foo(pydantic.BaseModel):
    a: str | None
    b: str | None

Foo(**{'a': 'a', 'b': None}).dict(exclude_unset=True)
# {'a': 'a', 'b': None}
```

File: backend/requirements.txt [hunk 8]
```diff
@@ -15,7 +15,7 @@ jinja2==3.1.2
 aiokafka==0.8.0
 uvloop==0.17.0
 nbformat>=5.4.0,<6
-deepchecks@git+https://github.com/deepchecks/deepchecks.git@main
+# deepchecks@git+https://github.com/deepchecks/deepchecks.git@main
```
Comment (yromanyshyn) [inline backend/requirements.txt:18]: did you forget to uncomment this line?

File: backend/tests/common.py [hunk 9]
```diff
@@ -735,11 +758,29 @@ def fetch_alert_rule(
 
         return self._assert_alert_rule(response.json())
 
+    def fetch_model_alert_rule(
```
Comment (yromanyshyn) [inline backend/tests/common.py:761]: similarly, maybe rename it to `fetch_data_ingestion_alert_rule`?

usual alert rules also exist within the context of a model, therefore this name is a little bit confusing and inconsistent with other parts of the code base

File: backend/deepchecks_monitoring/api/v1/configuration.py (hunk 1 above)
Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/configuration.py:248]: @lironhamra says it's better for our FE

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py (hunk 5 above)
Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py:71]: sure

File: backend/deepchecks_monitoring/api/v1/configuration.py (hunk 1 above)
Comment (yromanyshyn) [reply backend/deepchecks_monitoring/api/v1/configuration.py:248]: how exactly that will be better?

what I think is that the execution time of this endpoint will be pretty big/
we are executing here two complex queries sequentially

did you test it, what is the execution time of this endpoint currently?

Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/configuration.py:248]: pretty fast

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        data_ingestion_alert_rule_id: int,  # pylint: disable=unused-argument
+        resolved: t.Optional[bool] = None,
+        session: AsyncSession = AsyncSessionDep,
+        alert_rule: DataIngestionAlertRule = Depends(DataIngestionAlertRule.get_object_from_http_request),
+):
+    """Get list of alerts raised by a given data-ingestion alert rule."""
+    query = select(DataIngestionAlert).where(DataIngestionAlert.alert_rule_id == alert_rule.id)
+    if resolved is not None:
+        query = query.where(DataIngestionAlert.resolved.is_(resolved))
+    query = await session.execute(query.order_by(DataIngestionAlert.start_time))
```
[the remark this answers is not in the document]
Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py:144]: I think it might break current fe logic

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py (hunk 7 above)
Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py:115]: there are no defaults

Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py:115]: the body dict is read only I think

Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/data_ingestion_alert_rule.py:115]: It is like that in many endpoints

File: backend/deepchecks_monitoring/api/v1/data_ingestion_alert.py (hunk 4 above)
Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/data_ingestion_alert.py:74]: it is used in the fe like that

File: backend/deepchecks_monitoring/api/v1/configuration.py (hunk 2 above)
Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/configuration.py:242]: https://linear.app/deepchecks/issue/MON-2372/test-sql-execute-in-loops

File: backend/deepchecks_monitoring/api/v1/configuration.py (hunk 1 above)
Comment (JKL98ISR) [reply backend/deepchecks_monitoring/api/v1/configuration.py:248]: 100ms

**Review by matanper — APPROVED** [review]

**Review by matanper — APPROVED** [review]

**Review by matanper — APPROVED** [review]
